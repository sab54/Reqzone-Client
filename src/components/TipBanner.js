/**
 * TipBanner.js
 *
 * This component displays a simple banner with a lightbulb icon and a tip message.
 * It is used to surface helpful hints or guidance to the user inside the app.
 *
 * Key Features:
 * - **Highlight Background**: Uses a theme-provided `highlight` color (default: beige-like `#f5f5dc`) for the banner background.
 * - **Tip Text Styling**: Displays the provided `tip` text prefixed with a ðŸ’¡ emoji, styled using theme-provided `title` color (default: black).
 * - **Theming Support**: Accepts a `theme` prop to override colors for customization.
 *
 * Props:
 * - `tip` *(string, required)*: The tip message to display inside the banner.
 * - `theme` *(object, optional)*: Custom theme overrides:
 *   - `highlight` (string) â†’ background color of the banner.
 *   - `title` (string) â†’ text color of the tip.
 *
 * Accessibility:
 * - The entire banner is accessible as text content.
 *
 * Usage:
 * ```jsx
 * <TipBanner
 *   tip="Stay hydrated and take regular breaks!"
 *   theme={{ highlight: '#e0f7fa', title: '#00695c' }}
 * />
 * ```
 *
 * Author: Sunidhi Abhange
 */

import React from 'react';
import { View, Text, StyleSheet } from 'react-native';

const TipBanner = ({ tip, theme = {} }) => {
    const {
        highlight = '#f5f5dc', // default to a soft yellow-like highlight
        title = '#000', // default to black text
    } = theme;

    return (
        <View style={[styles.container, { backgroundColor: highlight }]}>
            <Text style={[styles.tipText, { color: title }]}>ðŸ’¡ {tip}</Text>
        </View>
    );
};

const styles = StyleSheet.create({
    container: {
        padding: 10,
        borderRadius: 8,
        marginBottom: 15,
    },
    tipText: {
        fontSize: 13,
        fontFamily: 'Poppins',
    },
});

export default TipBanner;
