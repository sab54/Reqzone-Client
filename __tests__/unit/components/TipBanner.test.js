/**
 * TipBanner.test.js
 *
 * What This Test File Covers:
 *
 * 1. Basic Rendering
 *    - Renders the tip text prefixed with ðŸ’¡.
 *
 * 2. Theming Support
 *    - Applies custom highlight and title colors from `theme`.
 *
 * 3. Defaults
 *    - Falls back to default highlight (`#f5f5dc`) and title (`#000`) when no theme provided.
 *
 * 4. Snapshot
 *    - Ensures the structure and text output remain stable.
 */

import React from 'react';
import { render } from '@testing-library/react-native';
import TipBanner from '../../../src/components/TipBanner';

const flatten = (style) =>
  Array.isArray(style) ? Object.assign({}, ...style) : style;

const findViewWithStyle = (node, predicate) => {
  if (!node || typeof node !== 'object') return null;
  const { type, props, children } = node;

  if (
    type === 'View' &&
    props &&
    props.style &&
    predicate(flatten(props.style))
  ) {
    return node;
  }
  if (Array.isArray(children)) {
    for (const c of children) {
      const found = findViewWithStyle(c, predicate);
      if (found) return found;
    }
  }
  return null;
};

describe('TipBanner', () => {
  test('renders tip text with ðŸ’¡ prefix', () => {
    const { getByText } = render(<TipBanner tip="Stay hydrated" />);
    expect(getByText(/ðŸ’¡\s*Stay hydrated/)).toBeTruthy();
  });

  test('applies theme highlight and title colors (no testID required)', () => {
    const theme = { highlight: '#abc123', title: '#654321' };
    const { getByText, toJSON } = render(
      <TipBanner tip="Custom tip" theme={theme} />
    );

    // Text color check
    const text = getByText(/ðŸ’¡\s*Custom tip/);
    const textStyle = flatten(text.props.style);
    expect(textStyle.color).toBe('#654321');

    // Container background via rendered tree
    const tree = toJSON();
    const container = findViewWithStyle(
      tree,
      (s) => s.backgroundColor === '#abc123'
    );
    expect(container).toBeTruthy();
  });

  test('uses default colors when no theme provided', () => {
    const { getByText, toJSON } = render(<TipBanner tip="Default theme tip" />);

    // Text default color
    const text = getByText(/ðŸ’¡\s*Default theme tip/);
    const textStyle = flatten(text.props.style);
    expect(textStyle.color).toBe('#000');

    // Background default color
    const tree = toJSON();
    const container = findViewWithStyle(
      tree,
      (s) => s.backgroundColor === '#f5f5dc'
    );
    expect(container).toBeTruthy();
  });

  test('matches snapshot', () => {
    const { toJSON } = render(<TipBanner tip="Snapshot tip" />);
    expect(toJSON()).toMatchSnapshot();
  });
});
