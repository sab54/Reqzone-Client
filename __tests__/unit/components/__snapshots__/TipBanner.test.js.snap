// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TipBanner matches snapshot 1`] = `
<View
  style={
    [
      {
        "borderRadius": 8,
        "marginBottom": 15,
        "padding": 10,
      },
      {
        "backgroundColor": "#f5f5dc",
      },
    ]
  }
>
  <Text
    style={
      [
        {
          "fontFamily": "Poppins",
          "fontSize": 13,
        },
        {
          "color": "#000",
        },
      ]
    }
  >
    ðŸ’¡ 
    Snapshot tip
  </Text>
</View>
`;
